### PR TITLE
Allowed multiple, general RZ coordinate systems

### LINKED ISSUE #24339: Add a generalized RZ coordinate system
## Reason
This would support a new coordinate system that generalizes the RZ coordinate system to not be restricted to the +X and +Y axes and starting at the origin.

## Design
To accomplish this, we can add a new coordinate system called `RZ_GENERAL` and have the user/developer supply the following:

- A list of blocks using this system
- A list of the axes' origins
- A list of the axes' directions

## Impact
New feature.

### PR DESCRIPTION
**Pull request by joshuahansel**
Closes #24339

I will soon add a link for a PR where this is adopted in THM.

### CODE REVIEW

**Comment by joshuahansel** [comment]
> what's the issue with RZ being RZ_GENERAL with a default axis ?

The main thing is that there is some functionality like transfers that I haven't done for RZ_GENERAL, and some applications might be relying on the old RZ stuff. A user might specify a general axis, but then the application is still getting the old RZ axis. Then I need to figure out if I can make everything backward compatible. There's the minor concern about the transformation factor computation being more expensive 🤷‍♂️ 

I'll explore this idea. I need to switch to some Sockeye for a bit.

**Comment by GiudGiud** [comment]
> The main thing is that there is some functionality like transfers that I haven't done for RZ_GENERAL, and some applications might be relying on the old RZ stuff.

This is all done in one class btw, in the https://github.com/idaholab/moose/blob/next/framework/src/utils/MooseAppCoordTransform.C

File: framework/src/base/Assembly.C [hunk 1]
```diff
@@ -58,11 +58,22 @@ coordTransformFactor(const MooseMesh & mesh,
...
+    if (mesh.usingGeneralAxisymmetricCoordAxes())
+    {
+      const auto & axis = mesh.getGeneralAxisymmetricCoordAxis(sub_id);
+      MooseMeshUtils::coordTransformFactorRZGeneral(point, axis, factor);
+    }
+    else
+    {
+      MooseMeshUtils::coordTransformFactor(
+          point, factor, coord_type, mesh.getAxisymmetricRadialCoord());
+    }
```
Comment (lindsayad) [inline framework/src/base/Assembly.C:73]:
```suggestion
      MooseMeshUtils::coordTransformFactor(
          point, factor, coord_type, mesh.getAxisymmetricRadialCoord());
```

File: framework/src/mesh/MooseMesh.C [hunk 2]
```diff
@@ -3646,6 +3681,70 @@ MooseMesh::setAxisymmetricCoordAxis(const MooseEnum & rz_coord_axis)
...
 }
 
+void
+MooseMesh::setGeneralAxisymmetricCoordAxes(
+    const std::vector<SubdomainName> & blocks,
+    const std::vector<std::pair<Point, RealVectorValue>> & axes)
+{
+  // Set the axes for the given blocks
+  mooseAssert(blocks.size() == axes.size(), "Blocks and axes vectors must be the same length.");
+  for (unsigned int i = 0; i < blocks.size(); ++i)
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:3691]:
```suggestion
  for (const auto i : index_range(blocks))
```

File: framework/src/mesh/MooseMesh.C [hunk 3]
```diff
@@ -3646,6 +3681,70 @@ MooseMesh::setAxisymmetricCoordAxis(const MooseEnum & rz_coord_axis)
...
+        mooseError("The block '",
+                   blocks[i],
+                   "' was provided in setGeneralAxisymmetricCoordAxes(), but the coordinate system "
+                   "for this block is not 'RZ'.");
+    }
+  }
+
+  // Make sure there are no RZ blocks that still do not have axes
+  const auto all_subdomain_ids = meshSubdomains();
+  for (const auto & subdomain_id : all_subdomain_ids)
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:3721]:
```suggestion
  for (const auto subdomain_id : all_subdomain_ids)
```

File: test/tests/coord_type/coord_type_rz_general.i [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+[Preconditioning]
+  [pc]
+    type = SMP
+    full = true
+  []
+[]
+
+[Executioner]
+  type = Transient
+
```
Comment (lindsayad) [inline test/tests/coord_type/coord_type_rz_general.i:357]:
```suggestion
```

File: test/tests/coord_type/tests [hunk 5]
```diff
@@ -32,4 +32,12 @@
...
     requirement = "The system shall support cylindrical coordinates with the radial direction in the 'y' direction with discontinous Galkerin and integrated boundary conditions."
   [../]
+  [rz_general]
+    type = Exodiff
+    input = coord_type_rz_general.i
+    exodiff = coord_type_rz_general.e
+    requirement = "The system shall support cylindrical coordinates with a general axis."
+    design = 'Mesh/index.md'
+    issues = '#24339'
+  []
```
Comment (lindsayad) [inline test/tests/coord_type/tests:42]: can we do a csv diff?

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/mesh/MooseMesh.C (hunk 3 above)
Comment (joshuahansel) [reply framework/src/mesh/MooseMesh.C:3721]: I guess this is just because making references to small data types is not advantageous?

File: test/tests/coord_type/tests (hunk 5 above)
Comment (joshuahansel) [reply test/tests/coord_type/tests:42]: Is it your philosophy to just do CSVDiff when Exodiff is not strictly necessary? In this particular case, I think an Exodiff is advantageous since we can check the whole distribution.

Comment (lindsayad) [reply test/tests/coord_type/tests:42]: In perusing it seemed like the postprocessors seemed like what you really cared about, but if you care about a field, then go with Exodiff!

Comment (joshuahansel) [reply test/tests/coord_type/tests:42]: Yeah that "T" little guy is in that show list, just hard to see.

**Review by lindsayad — APPROVED** [review]

**Comment by GiudGiud** [comment]
Awesome new feature, I like the documentation too!

**Comment by joshuahansel** [comment]
@GiudGiud Thanks for the suggestion to keep it in the `RZ` coordinate system - I'm pretty sure that was the right call.

**Comment by GiudGiud** [comment]
If I had to make another suggestion it would be to make a finite volume test. We rely on _coord iirc for the face area, so it may work right off the bat but maybe not

Also any kernel that says "RZ" could have some suspicious code to double check

**Comment by GiudGiud** [comment]
also update this page
https://mooseframework.inl.gov/source/base/RZSymmetry.html

**Comment by joshuahansel** [comment]
> If I had to make another suggestion it would be to make a finite volume test

My plan is to apply this to `FlowComponentNS` in the follow-up where I actually start using this in THM. Then this will have a test of FV in general RZ.
